ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision/BulkSync.hs

```diff
@@ -0,0 +1,551 @@
+) where
+
+import Control.Monad (guard)
+import Control.Monad.Class.MonadTime.SI (MonadMonotonicTime (getMonotonicTime), addTime)
+import Control.Monad.Trans.Maybe (MaybeT (MaybeT, runMaybeT))
+import Control.Monad.Writer.Strict (Writer, runWriter, MonadWriter (tell))
+import Control.Tracer (Tracer, traceWith)
+import Data.Bifunctor (first, Bifunctor (..))
+import Data.Function (on)
+import qualified Data.List as List
+import qualified Data.Set as Set
+import Data.Maybe (maybeToList)
+
+import Cardano.Prelude (partitionEithers)
+
+import Ouroboros.Network.AnchoredFragment (AnchoredFragment, headBlockNo)
+import qualified Ouroboros.Network.AnchoredFragment as AF
+import Ouroboros.Network.Block
+import Ouroboros.Network.BlockFetch.ClientState
+         (FetchRequest (..), PeersOrder (..), peerFetchBlocksInFlight)
+import Ouroboros.Network.BlockFetch.ConsensusInterface (FetchMode(FetchModeBulkSync))
+import Ouroboros.Network.BlockFetch.DeltaQ (calculatePeerFetchInFlightLimits)
+import Ouroboros.Network.BlockFetch.ConsensusInterface (ChainSelStarvation(..))
+
+import Ouroboros.Network.BlockFetch.Decision.Deadline
+import Ouroboros.Network.BlockFetch.Decision.Trace (TraceDecisionEvent (..))
+
+-- | A trivial foldable data structure with a 'Semigroup' instance that
+-- concatenates in @O(1)@. Only meant for short-term use, followed by one fold.
+data ListConcat a = List [a] | Concat (ListConcat a) (ListConcat a)
+
+instance Semigroup (ListConcat a) where
+  (<>) = Concat
+
+instance Monoid (ListConcat a) where
+  mempty = List []
+
+listConcatToList :: ListConcat a -> [a]
+listConcatToList = flip go []
+  where
+    go (List xs) acc = xs ++ acc
+    go (Concat x y) acc = go x (go y acc)
+
+type WithDeclined peer = Writer (ListConcat (FetchDecline, peer))
+
+runWithDeclined :: WithDeclined peer a -> (a, ListConcat (FetchDecline, peer))
+runWithDeclined = runWriter
+
+fetchDecisionsBulkSyncM
+  :: forall peer header block m extra.
+     (Ord peer,
+      HasHeader header,
+      HeaderHash header ~ HeaderHash block, MonadMonotonicTime m)
+  => Tracer m (TraceDecisionEvent peer header)
+  -> FetchDecisionPolicy header
+  -> AnchoredFragment header
+  -> (Point block -> Bool)
+  -> MaxSlotNo
+  -> ChainSelStarvation
+  -> ( PeersOrder peer
+     , PeersOrder peer -> m ()
+     , peer -> m ()
+     )
+  -> [(AnchoredFragment header, PeerInfo header peer extra)]
+  -> m [(FetchDecision (FetchRequest header), PeerInfo header peer extra)]
+fetchDecisionsBulkSyncM
+  tracer
+  fetchDecisionPolicy@FetchDecisionPolicy {bulkSyncGracePeriod}
+  currentChain
+  fetchedBlocks
+  fetchedMaxSlotNo
+  chainSelStarvation
+  ( peersOrder0,
+    writePeersOrder,
+    demoteCSJDynamo
+    )
+  candidatesAndPeers = do
+    let (peersOrder1, orderedCandidatesAndPeers) =
+          alignPeersOrderWithActualPeers
+            (peerInfoPeer . snd)
+            candidatesAndPeers
+            peersOrder0
+
+    peersOrder <- checkLastChainSelStarvation peersOrder1
```

Comment:
Here's a problem: `orderedCandidatesAndPeers` does not reflect the reordering that could happen in the later call to `checkLastChainSelStarvation`.

Flipping the order of the calls fixes that easily. Then new peers would be inserted behind the last starvation-producing peer, but I don't see now why it would be a problem.

Reply:
Addressed in 80e477a.